### DISCUSSION TITLE
How handle the connection errors?

### DISCUSSION THREAD

**Discussion by Pablete01** [supabase-community/supabase-kt#133] [Q&A]
Hi.
How handle the errors in case of bad internet connection or database error?
I have no problem when all is ok, but when I put some error in Supabase to test, the app crashes.
Can someboy tell me how handle with the errors?
I'm using Coroutines.

   private fun getClient(): SupabaseClient {
        return createSupabaseClient(
            "https://supabase","Key"
        ) {
            install(Postgrest)
        }
    }

and

  val supabaseResponse = client.postgrest.from("table")

  val data = supabaseResponse.decodeList<DataClass>()

Thanks.

**Comment by jan-tennert** [comment]
Should be in the documentation
![image](https://user-images.githubusercontent.com/26686035/236686222-cba171d6-62d3-40af-b9ff-99bd6426b575.png)

**Comment by jan-tennert** [comment]
You handle them with a try catch block / or using runCatching

**Comment by jan-tennert** [comment]
You can see the demo's viewmodel: https://github.com/supabase-community/supabase-kt/blob/master/demos/chat-demo-mpp/common/src/commonMain/kotlin/io/github/jan/supabase/common/ChatViewModel.kt
It always catches any error when doing a request

**Comment by sasa-boros** [comment]
Hello.

We are getting a connection error when we try to upload files to Supabase storage. It happens on the first upload and then happens again only after some idle time. Our app is running in a DEV environment and the weird thing is that the exception says "localhost". Do you have any idea what the problem can be?

The code in question:

```
val client = createSupabaseClient(
        supabaseUrl = supabaseUrl,
        supabaseKey = supabaseServiceKey)  {
            install(GoTrue) {
                requestTimeout = 60.seconds // Default: 10 seconds
            }
            install(Storage) {
                transferTimeout = 300.seconds // Default: 120 seconds
                requestTimeout = 120.seconds // Default: 10 seconds
            }
            //install other modules
        };

fun uploadFile(bucketId : String, filePath: String, fileContent: ByteArray) : String? {
        var upload: String?
        runBlocking {
            upload = client.storage[bucketId].upload(path = filePath, data = fileContent, upsert = true)
        }
        return upload
    }
```

When calling the upload function, the exception we get is the following:

```
at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104) ~[kotlinx-coroutines-core-jvm-1.6.4.jar!/:na]
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) ~[kotlin-stdlib-1.7.21.jar!/:1.7.21-release-272(1.7.21)]
at io.github.jan.supabase.network.KtorSupabaseHttpClient$request$1.invokeSuspend(KtorSupabaseHttpClient.kt) ~[supabase-kt-jvm-1.1.0.jar!/:na]
at io.github.jan.supabase.network.KtorSupabaseHttpClient.request(KtorSupabaseHttpClient.kt:55) ~[supabase-kt-jvm-1.1.0.jar!/:na]
io.github.jan.supabase.exceptions.HttpRequestException: HTTP request to http://localhost (PUT) failed with message: Connection reset

```

**Reply by jan-tennert** [reply]
Which Ktor Engine are you using?

**Reply by sasa-boros** [reply]
Thank you for your quick response! We are using Apache Engine:
```
<dependency>
	<groupId>io.ktor</groupId>
	<artifactId>ktor-client-apache-jvm</artifactId>
	<version>2.3.4</version>
</dependency>
```

**Reply by jan-tennert** [reply]
So I have seen this problem a few times, the triggers were mostly these
- missing internet intent
- the CIO engine was somehow broken
- the internet connection was bad
but I'm still not sure whats wrong with most of them including yours. I try to investigate further, can you share a reproducable example?

**Reply by sasa-boros** [reply]
Unfortunately, we are unable to reproduce it locally. It seems that it only happens when hosted (we are using AWS). We checked the Kotlin version and the version of all the other artifacts and it seems that all of them are the same as local, so no differences there. Some investigations suggested that it might be related to the TCP connection being broken (Connection: keep-alive is used). If we can help debug this anyhow, we will be happy to help out!

**Reply by jan-tennert** [reply]
The problem is this happens either sometimes/always or never, and it seems some engines are more affected than others. The Ktor client gets created [here](https://github.com/supabase-community/supabase-kt/blob/master/src/commonMain/kotlin/io/github/jan/supabase/network/KtorSupabaseHttpClient.kt).

**Reply by sasa-boros** [reply]
Yes, that is true, hard to find the source. In our case it doesn't work on the first try then it works without issues until you leave it idle, then it stops working again and the cycle repeats. My first idea was Spring bean context creation, then network issues, but in the end, nothing completely fits the problem.  For now, we added the retry mechanism in case of network failure i.e. on `HttpRequestException` exception, but I am hopeful we can solve it sometime in the future.
